backend/s3/file.go

```diff
@@ -287,48 +303,56 @@ func (f *File) getHeadObject() (*s3.HeadObjectOutput, error) {
 	return client.HeadObject(headObjectInput)
 }
 
-// Copy from S3-to-S3 when accounts are the same between source and target and fall back to a
-// TouchCopy() call if they are different.
-func (f *File) copyFile(targetFile *File) error {
+// For copy from S3-to-S3 when credentials are the same between source and target, return *s3.CopyObjectInput or error
+func (f *File) getCopyObjectInput(targetFile *File) (*s3.CopyObjectInput, error) {
+	//first we must determine if we're using the same s3 credentials for source and target before doing a native copy
 	isSameAccount := false
-	hasACLOption := false
+	var ACL string
 
-	opts, hasOptions := f.fileSystem.options.(Options)
-	if hasOptions {
-		hasACLOption = opts.ACL != ""
-	}
+	fileOptions := f.Location().FileSystem().(*FileSystem).options
+	targetOptions := targetFile.Location().FileSystem().(*FileSystem).options
 
-	if hasOptions && targetFile.fileSystem.options != nil {
-		isSameAccount = opts.AccessKeyID == targetFile.fileSystem.options.(Options).AccessKeyID
+	if fileOptions == nil && targetOptions == nil {
+		// if both opts are nil, we must be using the default credentials
+		isSameAccount = true
+	} else {
+		opts, hasOptions := fileOptions.(Options)
+		targetOpts, hasTargetOptions := targetOptions.(Options)
+		if hasOptions {
+			//use source ACL (even if empty), UNLESS target ACL is set
+			ACL = opts.ACL
+			if hasTargetOptions && targetOpts.ACL != "" {
+				ACL = targetOpts.ACL
+			}
+			if hasTargetOptions {
+				// since accesskey and session token are mutually exclusive, one will be nil
+				// if both are the same, we're using the same credentials
+				isSameAccount = (opts.AccessKeyID == targetOpts.AccessKeyID) && (opts.SessionToken == targetOpts.SessionToken)
+			}
+		}
 	}
 
 	// If both files use the same account, copy with native library. Otherwise, copy to disk
 	// first before pushing out to the target file's location.
-	copyInput := new(s3.CopyObjectInput).
-		SetKey(targetFile.key).
-		SetBucket(targetFile.bucket).
-		SetCopySource(path.Join(f.bucket, f.key))
-
-	if hasOptions && hasACLOption {
-		copyInput.SetACL(opts.ACL)
-	}
-
 	if isSameAccount {
-		client, err := f.fileSystem.Client()
+		key := strings.Replace(f.key, "%", "%25", -1)
```

Comment:
he's just encoding the '%' here. '%25' is '%'. '%20' is a space. 

This doesn't answer your questions though...

Reply:
So this only applies to the CopySource value which requires us to URL-encode all values.  Which we do when calling u.EscapedPath().  However when trying to copy the literal "some/file%20withspace.txt", EscapedPath() will not double-encode it.  So what happens is %20 gets interpolated the CopyObjectInput as a space.  Initially I regex'd only literal %20, but then I assumed we'd have the same effect if any other literal url-encoded character was saved to the source filesystem (like what happens to spaces when downloading from the GCS console UI).  
So here I'm assuming that all % signs found are literal and not URL-encoded characters. The backend integration test is what actually found this edge case.
 I'd prefer a consensus on whether this seems right.

Reply:
If it seems ok after my explanation, I can at least add comments to explain my decision in the code.  it's is a bit of "what-the-eff-is-this" code.

Reply:
Took me a bit to get my head wrapped around this.  It is a bit odd but I don't see a better way to do it.

Reply:
It looks like `u.EscapedPath()` and `url.PathEscape("foo/bar.txt")` follow different rules on what should be escaped. I tried `url.PathEscape()` and it will double encode your string properly.

https://play.golang.org/p/BnbIhX-JKqR

Reply:
Thanks Chris.  That's a much better solution. My tests still pass.